### ISSUE TITLE
multiple anomalies in coqchk on CompCert

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#3026]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 3026
Date: 2013-04-14 19:01:45 +0200
From: @_xavierleroy
To:   
CC:   @_barras, coq-bugs-redist@lists.gforge.inria.fr, @_pirbo, @_letouzey

Depends on:   #2588, #2933
See also:     https://coq.inria.fr/bugs/show_bug.cgi?id=3243
Last updated: 2017-06-07 14:36:29 +0200

## Bugzilla Comment ID: 5805
Date: 2013-04-14 19:01:45 +0200
From: @_xavierleroy

I have been trying for several years to run "coqchk" on CompCert, for additional confidence.  Below is the current status of my experiments, using the latest Coq release (8.4pl2) and the latest CompCert release (1.13).  There are still multiple "anomalies" and several definitions where coqchk just loops or takes forever.  

The existence of coqchk is an argument in favor of Coq for people who really care about high-assurance software.  I would therefore very much appreciate that you make it work properly.

- Xavier Leroy

How to reproduce:

wget http://compcert.inria.fr/release/compcert-1.13.tgz
tar xzf compcert-1.13.tgz
cd compcert-1.13
./configure ia32-linux
make -j <number of cores> proof

Edit the "check-proof:" entry at end of ./Makefile to add or remove -admit options as described below.  Then run "make check-proof".

1- Without any "-admit": loops or takes forever on
  checking cst: Integers.Int.Z_mod_modulus_range

2- With "-admit Integers": loops or takes forever on
  checking cst: Floats.Float.double_of_bits_of_double

3- With "-admit Integers -admit Floats": fails on
  checking ind: AST.external_function
Failure: impredicative Type inductive type. Please report.

4- With "-admit Integers -admit Floats -admit AST": fails on
  checking ind: Asm.instruction
Failure: impredicative Type inductive type. Please report.

5- With "-admit Integers -admit Floats -admit AST -admit Asm": fails on
  checking ind: Mach.instruction
Failure: impredicative Type inductive type. Please report.
Note that I reported this exact bug #2 years ago (PR#2588).

6- With "-admit Integers -admit Floats -admit AST -admit Asm -admit Mach": fails on
  checking cst: UnionFind.UF.elt
Anomaly: Uncaught exception Reduction.NotConvertible. Please report.

7- With "-admit Integers -admit Floats -admit AST -admit Asm -admit Mach -admit UnionFind": success! at last!

## Bugzilla Comment ID: 5840
Date: 2013-04-24 16:59:15 +0200
From: @_letouzey

Hi Xavier

I've started looking at these issues

1) Integers:
Unlike coqtop, the checker is also validating functor applications after expansion of the module parameters. So with these concrete parameters more reductions might occur, here for instance wordsize becomes convertible to 32%nat for Int, etc etc. I've a patch for giving a low priority during conversion to the expansion of constants coming from module parameter (see attachment), I'll see it this patch is ok with Bruno / Hugo.

2) Floats:
Not fully investigated yet, but I bet this one comes from the intensive use of vm_compute in this file. Since Coqchk uses compute instead of the vm, this might take ages. As a workaround, it seems that many computations in Floats could be replaced by some rewrites, this might help a bit (?). To be continued...

3,4,5) Cf #2588 indeed (and maybe #2933). I'm no expert in universes, I'll see with Bruno and/or Hugo.

6) UnionFind:
Another universe issue. There's something wrong with the Type typing elt in
(UNIONFIND with Definition elt := M.elt). Maybe a missing universe constraint somewhere ? This will also have to be investigated more deeply. Meantime, a possible workaround is to use <: instead of : in the definition of UF.
Separate issue: why this NotConvertible reach toplevel without being catched and turned into a nicer error message ?

Best,
Pierre L.

## Bugzilla Comment ID: 5841
Date: 2013-04-24 17:02:34 +0200
From: @_letouzey

Created attachment 400
Patch solving issue 1 : a minimal conv oracle for delaying unfold of module params

Patch for branch 8.4

> Attached file: 0001-Checker-conversion-heuristics-to-speed-up-validation.patch (text/plain, 4011 bytes)
> Description:   Patch solving issue 1 : a minimal conv oracle for delaying unfold of module params

## Bugzilla Comment ID: 5843
Date: 2013-04-24 23:42:24 +0200
From: @_letouzey

Bruno, what do you think of this patch about conversion ?

## Bugzilla Comment ID: 5847
Date: 2013-04-25 18:39:05 +0200
From: @_letouzey

Created attachment 401
Patch solving issue 6 : load inner mod constraints in coqchk

> Attached file: 0001-Checker-add-some-missing-inner-mod-constraints-part-.patch (text/plain, 1455 bytes)
> Description:   Patch solving issue 6 : load inner mod constraints in coqchk

## Bugzilla Comment ID: 5848
Date: 2013-04-25 18:42:40 +0200
From: @_letouzey

Concerning 6) (the NotConvertible in UnionFind), here is a minimal test-case:

Module Type FOO. Parameter t : Type. End FOO.
Module Type BAR. Parameter t : Type. End BAR.

Module M (X : BAR) : FOO with Definition t := X.t.
 Definition t := X.t.
End M.

It seems the universe constraints are simply not loaded by coqchk when inside a functor, see my second patch that appears to solve this... I wonder whether we should also add constraints appearing in module_type_body and in constant_body ?
For constant_body this used to be done but that was stopped by the fix for #1927...

One more time, I'll try to have feedback about this patch before inclusion...

Pierre

## Bugzilla Comment ID: 11180
Date: 2015-11-03 08:30:24 +0100
From: @_maximedenes

This patch doesn't seem necessary today (constraints carried by fields of a module are included in the ones of the module itself).

The minimal example extracted from UnionFind goes through since 8.4 pl4, probably due to Enrico's fix of #3243.

## Bugzilla Comment ID: 16647
Date: 2017-06-07 14:36:29 +0200
From: @_Zimmi48

From what I read here https://github.com/AbsInt/CompCert/blob/master/Makefile#L280-L285, coqchk still has issues with CompCert. Can this bug be updated to reflect the current limitations?
